### DISCUSSION TITLE
Does this help with simplifying P/Invoke with third-party .dll files?

### DISCUSSION THREAD

**Discussion by myokeeh** [microsoft/CsWin32#161] [Q&A] [answered]
I have a project that uses a third party .dll from a device manufacturer and they have provided a C# wrapper with a bunch of P/Invokes in it. Does this help in simplifying or getting rid of that wrapper code?

**Comment by AArnott** [comment, marked as answer]
No. CsWin32 only projects functions exported from Windows dll's -- not 3rd party ones.

Our technique however may be adopted by 3rd parties. The win32metadata repo first parses the header files with ClangSharp to produce a winmd file. Then CsWin32 transforms the winmd into C# source code.

**Reply by derek987654** [reply]
Does this get rid of the COM-nature of my library?  In other words, if I have a COM library and I utilize the method you describe above to access from C# source code, will I still have to Register my COM object on the target machine?

**Reply by AArnott** [reply]
@derek987654 AFAIK you only have to register COM objects that are *activatable* via CoCreateInstance. And you can do an "app local" registration to avoid machine-level impact. But I don't think that the code gen which produces COM interfaces has anything to do with the need (or lack thereof) to register anything.

**Reply by derek987654** [reply]
Thanks for the information Andrew - Is "app local" registration the same as "registrationless COM"  I'm aware of the latter, but my google searches didn't turn up anything on "app local."   If it is different, can you point me somewhere?  Thanks!

**Reply by AArnott** [reply]
Ah yes, "reg-free COM" was the term I was looking for. Not "app local". Sorry for the confusion.

**Reply by derek987654** [reply]
No worries.  Thanks again!
